app/shared/src/commonMain/kotlin/ui/subject/episode/EpisodeViewModel.kt

```diff
@@ -735,14 +753,60 @@ private class EpisodeViewModelImpl(
             }
         },
         onSend = { context, content ->
-            when (context) {
-                is CommentContext.Episode ->
-                    bangumiCommentService.postEpisodeComment(episodeId.value, content)
+            var turnstileResponse: CompletableDeferred<String>? = null
+            var triedCount = 0
+
+            coroutineScope {
+                val tokenCollectJob = launch {
+                    turnstileState.tokenFlow.collectLatest {
+                        turnstileResponse?.complete(it)
+                    }
+                }
+
+                launch {
+                    while (triedCount < 3) {
+                        withContext(Dispatchers.Main) { turnstileState.reload() }
+                        val turnstileToken = CompletableDeferred<String>()
+                            .also { turnstileResponse = it }
+                            .run { withTimeoutOrNull(10000L.milliseconds) { await() } }
+
+                        if (turnstileToken == null) {
+                            triedCount++
+                            continue
+                        }
+
+                        val response = when (context) {
+                            is CommentContext.Episode ->
+                                bangumiCommentService.postEpisodeComment(episodeId.value, content, turnstileToken)
 
-                is CommentContext.Reply ->
-                    bangumiCommentService.postEpisodeComment(episodeId.value, content, context.commentId)
+                            is CommentContext.Reply ->
+                                bangumiCommentService.postEpisodeComment(
+                                    episodeId.value,
+                                    content,
+                                    turnstileToken,
+                                    context.commentId,
+                                )
 
-                is CommentContext.SubjectReview -> {} // TODO: send subject comment
+                            is CommentContext.SubjectReview -> ApiResponse.success(Unit) // TODO: send subject comment
+                        }
+
+                        val error = response.failureOrNull()
+                        if (error == null) { // cancel all jobs in this scope to complete.
+                            tokenCollectJob.cancel()
+                            // 发送完了, 刷新一下
+                            commentStateRestarter.restart()
+                            withContext(Dispatchers.Main) { commentLazyListState.scrollToItem(0) }
```

Comment:
It's actually incorrect. We should start any UI calls from Compose `onClick` or so.  Try to make `CommentEditorState.send` suspend, then from UI you launch a coroutine in the ui scope:
```kotlin
val scope = rememberCoroutineScope()
Button({ scope.launch { sender.send();  commentLazyListState.scrollToItem(0) }})
```